### ISSUE TITLE
Consider stateful API/classes, i.e. scikit-learn estimators

### ISSUE THREAD

**Issue by seberg** [scientific-python/spatch#32]
Right now, we don't have any machinery for *stateful objects*.  The extreme case of this is scikit-learn estimators.
This isn't very important for users like SciPy who don't have much stateful API, but to push this towards scikit-learn, we would need a very clear story around this.

sklearn estimators have:
* An `__init__` which sets hyperparameters only.
* Typically a `fit()` that gets the arguments and creates internal state.
* Other methods that use or modify the internal state

Normalizing the state is plausible, but seems hard. Typically, once we picked an implementation, we have to stick to it for all other method calls, because otherwise we cannot guarantee compatibility.[^1]

To make matters even more confusing, one could do something like:
```
with backend_opts(...):
    est = Estimator(...)

est.fit()  # do we remember the options from when it was created or here?
```
(Not that I think the above is very worrying either way.)

This probably means that we need a concept of "stateful methods" for such classes. If one was called, we need to remember the backend and always re-use that as well pass in the state.[^2]
However, I need to have a clearer picture to really make up my mind about what the story is here!

There is also the component that estimators may actually make `.fit()` just call a single hidden `_fit_internal()` function to normalize things a bit first here.[^3]

---

The next piece of the puzzle, is that scikit-learn would want a way to convert from an estimator fitted in a backend to a "normal" one (and the opposite direction).
I expect this can be handled with two stateful-methods above.

[^1]: This is actually slightly different for estimators that are Array API enabled.  While their state differs in which types it uses, it always has the same structure.
[^2]: One could actually see the state as part of the supported types in theory. Then plugins could effectively indicate that they "understand" another implementations state object.
(We throw all types into one bin, but that bin could contain an estimator state object next to the arrays.)
[^3]: I would personally like to avoid this for simplicity on the scikit-learn side, but it might just be a lot clearer at least for certain estimators.

**Comment by eriknw** [comment]
Thanks for continuing to push on this and writing down your thoughts @seberg! You're right, this is interesting.

Here's an idea that continues towards the explicit dispatching w.r.t. state: specify attributes of an object in `@dispatchable`. For example, your last example could do
```python
    @dispatchable(["self._data", "x"])
    def __call__(self, x):
        # actual computation
```
which avoids a layer of indirection.

Heh, this implies a unary function can dispatch on multiple objects, such as:
```python
    @dispatchable(["self.left", "self.right"])
    def step(self):
        ...
```

This may be a little complex, since the function still receives `self` , so it may access other data on `self`. This may be positive or negative depending on situation. This also doesn't prevent the pattern you shared with passing explicit stateful objects.

**Comment by seberg** [comment]
Ah, that is an interesting trick when we have well defined state on the class, we could just say which of the objects on the class are part of that "state".
We would indeed still access all of these through the original `self`, I don't mind that, just means it is slightly fuzzier defined.
(For the SciPy JIT use-case, it probably you probably need the indirection to a staticmethod either way, but the backend can do that also.)

There are two cases here, and I think scikit-learn needs the other one, while SciPy may actually be happy with the first:
1. There is state, but the state is fully defined. In that case, defining which of the state is important (on the dispatchable or maybe also on the class) is actually quite nice.
2. It could be that `self.data` here is on opaque backend specific object.  In that case, no other backend can understand it.  There are actually two approaches in theory:
   * We bloat the type dispatching, and just give it a custom type which we add to the dispatching system.
   * We ensure that if we dispatched once, we stick to that backend.  So the `step` function will not dispatch again (it already knows which backend to choose).

I actually somewhat like dispatching every time, but overall, I lean towards it being impractical for complex state (but maybe not for SciPy!).

Thinking about it a bit yesterday, I was wondering about for the second case... And considering something in the direction of:
```
@stateful_class  # or subclass, etc.
class Estimator:
    @dispatchable(["x", "y"])
    def fit(self, x, y):
        pass
    @dispatched  # a method that doesn't make sense if not dispatched
    def other_func(self):
        pass
```
And then maybe the backend would be something like:
```
@implements_stateful_class
class EstimatorBackend:
    def __init__(self, ctx, original_estimator):
        # copy what we need from original_estimator?
        # or maybe `Estimator` can have a well defined interface
        # to call `EstimatorBackend._from_original(...)`
        # (This shouldn't have to be `__init__` to allow code-reuse
        # with user-facing code ideally, although if that is hard then nvm.)
    def fit(self, ctx, x, y):
        # backend fit.  But the __init__ will be called first
    def other_func(self, ctx):
        # might have ctx.original_estimator here also
        # context could be optional (as might allow code re-use)
        pass
```

**Comment by ev-br** [comment]
Thanks @seberg . Not that I understand details here, I'll just note that at least for SciPy it would be best if we could somehow avoid the dispatching work (and the corresponding perf penalty) on each `__call__` invocation. Currently, we can store the namespace on the instance and reuse it at each `__call__`.
